lib/dialects/postgres/connection-manager.js

```diff
@@ -18,40 +18,68 @@ class ConnectionManager extends AbstractConnectionManager {
     const pgLib = this._loadDialectModule('pg');
     this.lib = this.sequelize.config.native ? pgLib.native : pgLib;
 
+    this._clearDynamicOIDs();
     this._clearTypeParser();
     this.refreshTypeParser(dataTypes.postgres);
   }
 
   // Expose this as a method so that the parsing may be updated when the user has added additional, custom types
   _refreshTypeParser(dataType) {
-    if (dataType.types.postgres.oids) {
-      for (const oid of dataType.types.postgres.oids) {
-        this.oidMap[oid] = value => dataType.parse(value, oid, this.lib.types.getTypeParser);
+    const self = this;
+    const arrayParserBuilder = parser => {
+      return value => self.lib.types.arrayParser.create(value, parser).parse();
+    };
+    const rangeParserBuilder = parser => {
+      return value => dataType.parse(value, {parser});
+    };
+
+    // Set range parsers
+    if (dataType.key.toLowerCase() === 'range') {
+      for (const name in this.nameOidMap) {
+        const entry = this.nameOidMap[name];
+        if (! entry.rangeOid) continue;
+
+        const rangeParser = rangeParserBuilder(this.getTypeParser(entry.oid));
+        const arrayRangeParser = arrayParserBuilder(rangeParser);
+
+        this.oidParserMap.set(entry.rangeOid, rangeParser);
+        if (! entry.arrayRangeOid) continue;
+        this.oidParserMap.set(entry.arrayRangeOid, arrayRangeParser);
       }
+      return;
     }
 
-    if (dataType.types.postgres.array_oids) {
-      for (const oid of dataType.types.postgres.array_oids) {
-        this.arrayOidMap[oid] = value => {
-          return this.lib.types.arrayParser.create(value, v =>
-            dataType.parse(v, oid, this.lib.types.getTypeParser)
-          ).parse();
-        };
+    // Create parsers for normal or enum data types
+    const parser = value => dataType.parse(value);
+    const arrayParser = arrayParserBuilder(parser);
+
+    // Set enum parsers
+    if (dataType.key.toLowerCase() === 'enum') {
+      for (const oid of this.enumOids.oids || []) {
+        this.oidParserMap.set(oid, parser);
+      }
+      for (const arrayOid of this.enumOids.arrayOids || []) {
+        this.oidParserMap.set(arrayOid, arrayParser);
       }
+      return;
+    }
+
+    // Set parsers for normal data types
+    for (const name of dataType.types.postgres) {
+      if (! this.nameOidMap[name]) continue;
+      this.oidParserMap.set(this.nameOidMap[name].oid, parser);
+
+      if (! this.nameOidMap[name].arrayOid) continue;
+      this.oidParserMap.set(this.nameOidMap[name].arrayOid, arrayParser);
     }
   }
 
   _clearTypeParser() {
```

Comment:
`_clearTypeParser` should call `_clearDynamicOIDs`

Reply:
Every invocation of `_clearTypeParser` is also preceded by `_clearDynamicOIDs` (at least in connection manager) so that method should actually take care of clearing dynamic oids as well

But I understand your hesitation, they both are doing different things. You can leave this for now.